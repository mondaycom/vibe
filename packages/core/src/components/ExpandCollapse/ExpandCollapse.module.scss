@import "~monday-ui-style/dist/mixins";

.expandCollapse {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  color: var(--primary-text-color);
  background-color: transparent;
  box-sizing: border-box;
  border-radius: var(--border-radius-small);
}

.header {
  @include reset-button();
  padding: var(--spacing-medium);
  display: flex;
  justify-content: space-between;
  align-items: center;
  cursor: pointer;
}

.header:hover {
  background-color: var(--primary-background-hover-color);
}

.header:active {
  transform: scale(1);
}

.header .headerContent {
  margin: 0;
}

.header .iconComponent {
  color: var(--icon-color);
}

.content {
  padding: var(--spacing-medium);
}

.section {
  width: 100%;
}

.hideBorder {
  border: none;
}

.hideBorderBottom {
  border-bottom: none;
}

.showBorder {
  border: 1px solid var(--ui-border-color);
  .headerOpen {
    border-bottom: 1px solid;
    border-color: var(--ui-border-color);
  }
}

.animateIconOpen {
  transform: rotate(-180deg);
  transition: transform var(--motion-expressive-short) var(--motion-timing-transition);
}

.animateIconClose {
  transform: rotate(-360deg);
  transition: transform var(--motion-expressive-short) var(--motion-timing-transition);
}

.header.leftIcon {
  justify-content: flex-start;
  gap: var(--spacing-small);
}
