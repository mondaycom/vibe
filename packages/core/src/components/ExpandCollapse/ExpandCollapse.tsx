import cx from "classnames";
import React, { FC, forwardRef, ReactElement, useCallback, useRef, useState } from "react";
import useMergeRef from "../../hooks/useMergeRef";
import Icon from "../Icon/Icon";
import Text from "../Text/Text";
import { DropdownChevronDown } from "@vibe/icons";
import { VibeComponentProps, ElementContent } from "../../types";
import { ComponentDefaultTestId, getTestId } from "../../tests/test-ids-utils";
import styles from "./ExpandCollapse.module.scss";

export interface ExpandCollapseProps extends VibeComponentProps {
  /**
   * Custom renderer for the header component.
   */
  headerComponentRenderer?: () => ReactElement;
  /**
   * Class name applied to the header.
   */
  headerClassName?: string;
  /**
   * Class name applied to the content.
   */
  contentClassName?: string;
  /**
   * Class name applied to the root component.
   */
  componentClassName?: string;
  /**
   * The title displayed in the header.
   */
  title?: ElementContent;
  /**
   * The content inside the expandable section.
   */
  children?: ElementContent;
  /**
   * The size of the expand/collapse icon.
   */
  iconSize?: number | string;
  /**
   * If true, the section is open by default when rendered.
   */
  defaultOpenState?: boolean;
  /**
   * Controls the open state of the section.
   */
  open?: boolean;
  /**
   * Callback fired when the header is clicked.
   */
  onClick?: (event: React.MouseEvent) => void;
  /**
   * If true, hides the border around the component.
   */
  hideBorder?: boolean;
  /**
   * If true, captures the click event on the button.
   */
  captureOnClick?: boolean;
}

const ExpandCollapse: FC<ExpandCollapseProps> = forwardRef(
  (
    {
      children,
      headerComponentRenderer = null,
      title = "",
      className,
      defaultOpenState = false,
      iconSize = 24,
      id = "",
      open,
      onClick = null,
      hideBorder = false,
      headerClassName,
      contentClassName,
      componentClassName,
      "data-testid": dataTestId,
      captureOnClick = true
    },
    ref
  ) => {
    const componentRef = useRef(null);
    const mergedRef = useMergeRef(ref, componentRef);

    const [isOpen, setIsOpen] = useState(defaultOpenState);
    const isExpanded = open === undefined ? isOpen : open;

    const toggleExpand = () => {
      setIsOpen(!isOpen);
    };
    const renderHeader = useCallback(() => {
      return typeof title === "string" ? (
        <Text type="text1" className={cx(styles.headerContent)}>
          {title}
        </Text>
      ) : (
        title
      );
    }, [title]);

    return (
      <div
        ref={mergedRef}
        className={className}
        id={id}
        data-testid={dataTestId || getTestId(ComponentDefaultTestId.EXPAND_COLLAPSE, id)}
      >
        <div
          className={cx(
            styles.expandCollapse,
            {
              [styles.hideBorder]: hideBorder,
              [styles.showBorder]: !hideBorder
            },
            componentClassName
          )}
        >
          <button
            type="button"
            className={cx(styles.header, styles.section, headerClassName, {
              [styles.headerOpen]: isExpanded,
              [styles.hideBorderBottom]: hideBorder
            })}
            onClickCapture={captureOnClick ? onClick || toggleExpand : undefined}
            onClick={!captureOnClick ? onClick || toggleExpand : undefined}
            aria-expanded={isExpanded}
            aria-controls={`${id}-controls`}
          >
            {typeof title !== "string" || title.length !== 0
              ? renderHeader()
              : headerComponentRenderer && headerComponentRenderer()}
            <Icon
              className={cx(styles.iconComponent, {
                [styles.animateIconOpen]: isExpanded,
                [styles.animateIconClose]: !isExpanded
              })}
              iconType="svg"
              icon={DropdownChevronDown}
              iconSize={iconSize}
            />
          </button>
          {isExpanded && (
            <div
              className={cx(styles.content, styles.section, contentClassName, {
                [styles.animateExpandCollapseContent]: isExpanded
              })}
              id={`${id}-controls`}
              role="region"
            >
              {children}
            </div>
          )}
        </div>
      </div>
    );
  }
);

export default ExpandCollapse;
