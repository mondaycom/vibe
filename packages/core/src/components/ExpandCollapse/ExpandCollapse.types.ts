export type ExpandCollapseIconPosition = "left" | "right";
