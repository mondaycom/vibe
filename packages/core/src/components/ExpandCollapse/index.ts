export { default as ExpandCollapse, type ExpandCollapseProps } from "./ExpandCollapse";
