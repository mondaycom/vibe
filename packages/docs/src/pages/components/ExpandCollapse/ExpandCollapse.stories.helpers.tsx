import React from "react";
import { StorybookLink, Tip } from "vibe-storybook-components";

export const TipCombineMultiple = () => (
  <Tip title="Want to combine multiple ExpandCollapse?">
    Use our{" "}
    <StorybookLink size={StorybookLink.sizes.SMALL} page="Components/Accordion">
      Accordion
    </StorybookLink>{" "}
    component
  </Tip>
);
