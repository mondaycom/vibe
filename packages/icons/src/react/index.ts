/* eslint-disable */
/* tslint:disable */
export {default as Academy} from './Academy';
export {default as Activity} from './Activity';
export {default as Add} from './Add';
export {default as AddNewDoc} from './AddNewDoc';
export {default as AddSmall} from './AddSmall';
export {default as AddToTeam} from './AddToTeam';
export {default as AddUpdate} from './AddUpdate';
export {default as Alert} from './Alert';
export {default as AlignCenter} from './AlignCenter';
export {default as AlignLeft} from './AlignLeft';
export {default as AlignRight} from './AlignRight';
export {default as Announcement} from './Announcement';
export {default as API} from './API';
export {default as Apps} from './Apps';
export {default as Archive} from './Archive';
export {default as Attach} from './Attach';
export {default as Baseline} from './Baseline';
export {default as Basic} from './Basic';
export {default as Battery} from './Battery';
export {default as BlockQuote} from './BlockQuote';
export {default as Board} from './Board';
export {default as BoardPrivate} from './BoardPrivate';
export {default as BoardShareable} from './BoardShareable';
export {default as BoardTemplate} from './BoardTemplate';
export {default as Bold} from './Bold';
export {default as Bolt} from './Bolt';
export {default as Bookmark} from './Bookmark';
export {default as Broadcast} from './Broadcast';
export {default as Broom} from './Broom';
export {default as Bug} from './Bug';
export {default as Bullet} from './Bullet';
export {default as Bullets} from './Bullets';
export {default as Bulllet} from './Bulllet';
export {default as Calendar} from './Calendar';
export {default as Chart} from './Chart';
export {default as Check} from './Check';
export {default as Checkbox} from './Checkbox';
export {default as CheckList} from './CheckList';
export {default as ClassicFolder} from './ClassicFolder';
export {default as Clear} from './Clear';
export {default as Clipboard} from './Clipboard';
export {default as Close} from './Close';
export {default as CloseMedium} from './CloseMedium';
export {default as CloseRound} from './CloseRound';
export {default as CloseSmall} from './CloseSmall';
export {default as Code} from './Code';
export {default as Collapse} from './Collapse';
export {default as CollapseHorizontally} from './CollapseHorizontally';
export {default as CollapseRound} from './CollapseRound';
export {default as Column} from './Column';
export {default as Comment} from './Comment';
export {default as Completed} from './Completed';
export {default as Connect} from './Connect';
export {default as ConnectedDoc} from './ConnectedDoc';
export {default as ContentDirectory} from './ContentDirectory';
export {default as ConvertToItem} from './ConvertToItem';
export {default as ConvertToSubitem} from './ConvertToSubitem';
export {default as Counter} from './Counter';
export {default as Country} from './Country';
export {default as CreditCard} from './CreditCard';
export {default as Custom} from './Custom';
export {default as Cut} from './Cut';
export {default as Dashboard} from './Dashboard';
export {default as DashboardPrivate} from './DashboardPrivate';
export {default as Deactivate} from './Deactivate';
export {default as Delete} from './Delete';
export {default as Dependency} from './Dependency';
export {default as Description} from './Description';
export {default as DisabledUser} from './DisabledUser';
export {default as Divider} from './Divider';
export {default as Doc} from './Doc';
export {default as DocPrivate} from './DocPrivate';
export {default as DocShareable} from './DocShareable';
export {default as DocTemplate} from './DocTemplate';
export {default as DoubleCheck} from './DoubleCheck';
export {default as Downgrade} from './Downgrade';
export {default as Download} from './Download';
export {default as Drag} from './Drag';
export {default as Dropdown} from './Dropdown';
export {default as DropdownChevronDown} from './DropdownChevronDown';
export {default as DropdownChevronLeft} from './DropdownChevronLeft';
export {default as DropdownChevronRight} from './DropdownChevronRight';
export {default as DropdownChevronUp} from './DropdownChevronUp';
export {default as DueDate} from './DueDate';
export {default as Duplicate} from './Duplicate';
export {default as Edit} from './Edit';
export {default as Education} from './Education';
export {default as Email} from './Email';
export {default as Embed} from './Embed';
export {default as Emoji} from './Emoji';
export {default as Enter} from './Enter';
export {default as Erase} from './Erase';
export {default as Event} from './Event';
export {default as Expand} from './Expand';
export {default as ExternalPage} from './ExternalPage';
export {default as Favorite} from './Favorite';
export {default as Feedback} from './Feedback';
export {default as File} from './File';
export {default as Files} from './Files';
export {default as Filter} from './Filter';
export {default as Folder} from './Folder';
export {default as Form} from './Form';
export {default as Formula} from './Formula';
export {default as Forum} from './Forum';
export {default as Forward} from './Forward';
export {default as Fullscreen} from './Fullscreen';
export {default as FullscreenClose} from './FullscreenClose';
export {default as Gallery} from './Gallery';
export {default as Gantt} from './Gantt';
export {default as Gif} from './Gif';
export {default as Globe} from './Globe';
export {default as Graph} from './Graph';
export {default as Group} from './Group';
export {default as Guest} from './Guest';
export {default as Health} from './Health';
export {default as Heart} from './Heart';
export {default as Help} from './Help';
export {default as Hide} from './Hide';
export {default as Highlight} from './Highlight';
export {default as HighlightColorBucket} from './HighlightColorBucket';
export {default as Home} from './Home';
export {default as Idea} from './Idea';
export {default as Image} from './Image';
export {default as Inbox} from './Inbox';
export {default as Info} from './Info';
export {default as Integrations} from './Integrations';
export {default as Invite} from './Invite';
export {default as IPRestrictions} from './IPRestrictions';
export {default as Italic} from './Italic';
export {default as Item} from './Item';
export {default as ItemDefaultValues} from './ItemDefaultValues';
export {default as ItemHeightDouble} from './ItemHeightDouble';
export {default as ItemHeightSingle} from './ItemHeightSingle';
export {default as ItemHeightTriple} from './ItemHeightTriple';
export {default as Key} from './Key';
export {default as Keyboard} from './Keyboard';
export {default as Labs} from './Labs';
export {default as Launch} from './Launch';
export {default as Layout} from './Layout';
export {default as LearnMore} from './LearnMore';
export {default as Lines} from './Lines';
export {default as Link} from './Link';
export {default as Location} from './Location';
export {default as Locked} from './Locked';
export {default as LogIn} from './LogIn';
export {default as LogOut} from './LogOut';
export {default as LongText} from './LongText';
export {default as Mention} from './Mention';
export {default as Menu} from './Menu';
export {default as Microphone} from './Microphone';
export {default as Minimize} from './Minimize';
export {default as MinusSmall} from './MinusSmall';
export {default as Mirror} from './Mirror';
export {default as Mobile} from './Mobile';
export {default as MondayDoc} from './MondayDoc';
export {default as MondayLogoOutline} from './MondayLogoOutline';
export {default as Moon} from './Moon';
export {default as MoreActions} from './MoreActions';
export {default as MoreBelow} from './MoreBelow';
export {default as MoreBelowFilled} from './MoreBelowFilled';
export {default as Move} from './Move';
export {default as MoveArrowDown} from './MoveArrowDown';
export {default as MoveArrowLeft} from './MoveArrowLeft';
export {default as MoveArrowLeftDouble} from './MoveArrowLeftDouble';
export {default as MoveArrowLeftNarrow} from './MoveArrowLeftNarrow';
export {default as MoveArrowRight} from './MoveArrowRight';
export {default as MoveArrowRightNarrow} from './MoveArrowRightNarrow';
export {default as MoveArrowUp} from './MoveArrowUp';
export {default as Mute} from './Mute';
export {default as MyWeek} from './MyWeek';
export {default as NavigationChevronDown} from './NavigationChevronDown';
export {default as NavigationChevronLeft} from './NavigationChevronLeft';
export {default as NavigationChevronRight} from './NavigationChevronRight';
export {default as NavigationChevronUp} from './NavigationChevronUp';
export {default as NavigationDoubleChevronLeft} from './NavigationDoubleChevronLeft';
export {default as NewTab} from './NewTab';
export {default as Night} from './Night';
export {default as NoColor} from './NoColor';
export {default as Note} from './Note';
export {default as Notifications} from './Notifications';
export {default as NotificationsMuted} from './NotificationsMuted';
export {default as Numbers} from './Numbers';
export {default as Offline} from './Offline';
export {default as Open} from './Open';
export {default as Page} from './Page';
export {default as Paste} from './Paste';
export {default as Pause} from './Pause';
export {default as PDF} from './PDF';
export {default as Person} from './Person';
export {default as PersonRound} from './PersonRound';
export {default as Pin} from './Pin';
export {default as PinFull} from './PinFull';
export {default as Placeholder} from './Placeholder';
export {default as Play} from './Play';
export {default as Print} from './Print';
export {default as Prompt} from './Prompt';
export {default as PushNotification} from './PushNotification';
export {default as Quote} from './Quote';
export {default as Radio} from './Radio';
export {default as Recurring} from './Recurring';
export {default as RecycleBin} from './RecycleBin';
export {default as Redo} from './Redo';
export {default as Remove} from './Remove';
export {default as Reply} from './Reply';
export {default as ReplyAll} from './ReplyAll';
export {default as Retry} from './Retry';
export {default as Robot} from './Robot';
export {default as Rotate} from './Rotate';
export {default as ScheduledEmail} from './ScheduledEmail';
export {default as Search} from './Search';
export {default as Security} from './Security';
export {default as Send} from './Send';
export {default as Settings} from './Settings';
export {default as SettingsKnobs} from './SettingsKnobs';
export {default as Share} from './Share';
export {default as ShortText} from './ShortText';
export {default as Show} from './Show';
export {default as Shredder} from './Shredder';
export {default as Shuffle} from './Shuffle';
export {default as Sort} from './Sort';
export {default as SortAscending} from './SortAscending';
export {default as SortDescending} from './SortDescending';
export {default as Sound} from './Sound';
export {default as Status} from './Status';
export {default as StrikethroughS} from './StrikethroughS';
export {default as StrikethroughT} from './StrikethroughT';
export {default as Subitems} from './Subitems';
export {default as Sun} from './Sun';
export {default as Switch} from './Switch';
export {default as Switcher} from './Switcher';
export {default as Table} from './Table';
export {default as Tags} from './Tags';
export {default as Team} from './Team';
export {default as Text} from './Text';
export {default as TextBig} from './TextBig';
export {default as Textcolor} from './Textcolor';
export {default as TextColorIndicator} from './TextColorIndicator';
export {default as TextCopy} from './TextCopy';
export {default as TextHuge} from './TextHuge';
export {default as TextMedium} from './TextMedium';
export {default as TextSmall} from './TextSmall';
export {default as ThumbsDown} from './ThumbsDown';
export {default as ThumbsUp} from './ThumbsUp';
export {default as Time} from './Time';
export {default as Timeline} from './Timeline';
export {default as Translation} from './Translation';
export {default as TurnInto} from './TurnInto';
export {default as Underline} from './Underline';
export {default as Undo} from './Undo';
export {default as Unlocked} from './Unlocked';
export {default as Update} from './Update';
export {default as Upgrade} from './Upgrade';
export {default as Upload} from './Upload';
export {default as UserDomain} from './UserDomain';
export {default as UserStatus} from './UserStatus';
export {default as Versioning} from './Versioning';
export {default as Video} from './Video';
export {default as Wand} from './Wand';
export {default as Warning} from './Warning';
export {default as WarningFull} from './WarningFull';
export {default as WhatsNew} from './WhatsNew';
export {default as Widgets} from './Widgets';
export {default as Work} from './Work';
export {default as Workflow} from './Workflow';
export {default as Workspace} from './Workspace';
export {default as WrapText} from './WrapText';
/* tslint:enable */
/* eslint-enable */
