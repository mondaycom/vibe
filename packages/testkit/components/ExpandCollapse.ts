import { Page, Locator, test } from "@playwright/test";
import { BaseElement } from "./BaseElement";

/**
 * Class representing an ExpandCollapse element.
 * Extends the BaseElement class.
 */
export class ExpandCollapse extends BaseElement {
  private headerButtonLocator: Locator;
  private contentLocator: Locator;

  /**
   * Create an ExpandCollapse.
   * @param {Page} page - The Playwright page object.
   * @param {Locator} locator - The locator for the ExpandCollapse element.
   * @param {string} elementReportName - The name for reporting purposes.
   */
  constructor(page: Page, locator: Locator, elementReportName: string) {
    super(page, locator, elementReportName);
    this.headerButtonLocator = locator.locator("button[aria-expanded]");
    this.contentLocator = locator.locator("[role='region']");
  }

  /**
   * Click the header to toggle expand/collapse state.
   * @returns {Promise<void>}
   */
  async toggle(): Promise<void> {
    await test.step(`Toggle ${this.elementReportName}`, async () => {
      await this.headerButtonLocator.click();
    });
  }

  /**
   * Expand the component if it's currently collapsed.
   * @returns {Promise<void>}
   */
  async expand(): Promise<void> {
    await test.step(`Expand ${this.elementReportName}`, async () => {
      const isExpanded = await this.isExpanded();
      if (!isExpanded) {
        await this.headerButtonLocator.click();
      }
    });
  }

  /**
   * Collapse the component if it's currently expanded.
   * @returns {Promise<void>}
   */
  async collapse(): Promise<void> {
    await test.step(`Collapse ${this.elementReportName}`, async () => {
      const isExpanded = await this.isExpanded();
      if (isExpanded) {
        await this.headerButtonLocator.click();
      }
    });
  }

  /**
   * Check if the component is expanded.
   * @returns {Promise<boolean>} True if expanded, false if collapsed.
   */
  async isExpanded(): Promise<boolean> {
    let isExpandedAttribute: string | null = null;

    await test.step(`Check if ${this.elementReportName} is expanded`, async () => {
      isExpandedAttribute = await this.headerButtonLocator.getAttribute("aria-expanded", { timeout: 30000 });

      if (isExpandedAttribute === null) {
        throw new Error(`Attribute aria-expanded did not exist after 30000 ms`);
      }
    });

    return isExpandedAttribute === "true";
  }

  /**
   * Check if the component is collapsed.
   * @returns {Promise<boolean>} True if collapsed, false if expanded.
   */
  async isCollapsed(): Promise<boolean> {
    const expanded = await this.isExpanded();
    return !expanded;
  }

  /**
   * Get the header/title text.
   * @returns {Promise<string>} The header text content.
   */
  async getHeaderText(): Promise<string> {
    let headerText = "";
    await test.step(`Get header text of ${this.elementReportName}`, async () => {
      headerText = await this.headerButtonLocator.innerText();
    });
    return headerText;
  }

  /**
   * Check if the content section is visible.
   * @returns {Promise<boolean>} True if content is visible, false otherwise.
   */
  async isContentVisible(): Promise<boolean> {
    let isVisible = false;
    await test.step(`Check if content of ${this.elementReportName} is visible`, async () => {
      isVisible = await this.contentLocator.isVisible();
    });
    return isVisible;
  }

  /**
   * Get the content element for further interactions.
   * @returns {BaseElement} The content element wrapped in BaseElement.
   */
  getContentElement(): BaseElement {
    return new BaseElement(this.page, this.contentLocator, `${this.elementReportName} content`);
  }

  /**
   * Get the content text.
   * @returns {Promise<string>} The content text.
   */
  async getContentText(): Promise<string> {
    let contentText = "";
    await test.step(`Get content text of ${this.elementReportName}`, async () => {
      // Wait for content to be visible before getting text
      await this.contentLocator.waitFor({ state: "visible" });
      contentText = await this.contentLocator.innerText();
    });
    return contentText;
  }

  /**
   * Wait for the component to be expanded.
   * @returns {Promise<void>}
   */
  async waitForExpanded(): Promise<void> {
    await test.step(`Wait for ${this.elementReportName} to be expanded`, async () => {
      await this.headerButtonLocator.waitFor({ state: "visible" });
      // Poll until the component is expanded
      while (!(await this.isExpanded())) {
        await this.page.waitForTimeout(100);
      }
    });
  }

  /**
   * Wait for the component to be collapsed.
   * @returns {Promise<void>}
   */
  async waitForCollapsed(): Promise<void> {
    await test.step(`Wait for ${this.elementReportName} to be collapsed`, async () => {
      await this.headerButtonLocator.waitFor({ state: "visible" });
      // Poll until the component is collapsed
      while (await this.isExpanded()) {
        await this.page.waitForTimeout(100);
      }
    });
  }

  /**
   * Click on an item within the content area by its text.
   * @param {string} itemText - The text of the item to click.
   * @returns {Promise<void>}
   */
  async clickItemByText(itemText: string): Promise<void> {
    await test.step(`Click item "${itemText}" in ${this.elementReportName} content`, async () => {
      // Ensure the component is expanded first
      await this.expand();

      // Wait for content to be visible
      await this.contentLocator.waitFor({ state: "visible" });

      // Find and click the item
      const itemLocator = this.contentLocator.getByText(itemText, { exact: false });
      await itemLocator.click();
    });
  }

  /**
   * Click on an item within the content area by exact text match.
   * @param {string} itemText - The exact text of the item to click.
   * @returns {Promise<void>}
   */
  async clickItemByExactText(itemText: string): Promise<void> {
    await test.step(`Click item with exact text "${itemText}" in ${this.elementReportName} content`, async () => {
      // Ensure the component is expanded first
      await this.expand();

      // Wait for content to be visible
      await this.contentLocator.waitFor({ state: "visible" });

      // Find and click the item with exact text match
      const itemLocator = this.contentLocator.getByText(itemText, { exact: true });
      await itemLocator.click();
    });
  }

  /**
   * Click on a link within the content area by its text.
   * @param {string} linkText - The text of the link to click.
   * @returns {Promise<void>}
   */
  async clickLinkInContent(linkText: string): Promise<void> {
    await test.step(`Click link "${linkText}" in ${this.elementReportName} content`, async () => {
      // Ensure the component is expanded first
      await this.expand();

      // Wait for content to be visible
      await this.contentLocator.waitFor({ state: "visible" });

      // Find and click the link
      const linkLocator = this.contentLocator.locator("a").filter({ hasText: linkText });
      await linkLocator.click();
    });
  }
}
