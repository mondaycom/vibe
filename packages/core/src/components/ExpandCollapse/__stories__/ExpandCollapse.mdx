import { Meta } from "@storybook/blocks";
import { ACCORDION, DROPDOWN, TABLE } from "../../../storybook/components/related-components/component-description-map";
import { TipCombineMultiple } from "./ExpandCollapse.stories.helpers";
import * as ExpandCollapseStories from "./ExpandCollapse.stories";

<Meta of={ExpandCollapseStories} />

# ExpandCollapse

- [Overview](#overview)
- [Props](#props)
- [Variants](#variants)
- [Related components](#related-components)
- [Feedback](#feedback)

## Overview

ExpandCollapse is a component that allows you to hide and show content.

<Canvas of={ExpandCollapseStories.Overview} />

### Import

```js
import { ExpandCollapse } from "@vibe/core";
```

<TipCombineMultiple />

## Props

<PropsTable />

## Variants

### Open by default

<Canvas of={ExpandCollapseStories.OpenByDefault} />

### Controlled open state

You can control the open state of the ExpandCollapse component by passing the <code>open</code> prop.

<Canvas of={ExpandCollapseStories.ControlledOpenState} />

### Custom header renderer

<Canvas of={ExpandCollapseStories.CustomHeaderRenderer} />

### Without borders

<Canvas of={ExpandCollapseStories.WithoutBorders} />

## Related components

<RelatedComponents componentsNames={[ACCORDION, DROPDOWN, TABLE]} />
