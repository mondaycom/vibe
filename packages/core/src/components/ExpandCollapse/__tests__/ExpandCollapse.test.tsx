import { describe, it, expect } from "vitest";
import React from "react";
import renderer from "react-test-renderer";
import { render, fireEvent, screen } from "@testing-library/react";
import ExpandCollapse from "../ExpandCollapse";

describe("ExpandCollapse", () => {
  it("renders correctly with empty props", () => {
    const tree = renderer.create(<ExpandCollapse />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Should render header component", () => {
    render(<ExpandCollapse headerComponentRenderer={() => <h1>Some Header</h1>} />);

    expect(screen.getByText("Some Header")).toBeInTheDocument();
  });

  it("Should not render child components when closed", () => {
    render(
      <ExpandCollapse headerComponentRenderer={() => <h1>Some Header</h1>}>
        <h1>Child 1</h1>
        <span>Child 2</span>
      </ExpandCollapse>
    );

    expect(screen.getByText("Some Header")).toBeInTheDocument();

    expect(screen.queryByText("Child 1")).toBeNull();
    expect(screen.queryByText("Child 2")).toBeNull();
  });

  it("Should render header and child components when defaultOpenState = true", () => {
    render(
      <ExpandCollapse defaultOpenState={true} headerComponentRenderer={() => <h1>Some Header</h1>}>
        <h1>Child 1</h1>
        <span>Child 2</span>
      </ExpandCollapse>
    );

    expect(screen.getByText("Some Header")).toBeInTheDocument();

    expect(screen.getByText("Child 1")).toBeInTheDocument();
    expect(screen.getByText("Child 2")).toBeInTheDocument();
  });

  it("Should render child components after clicking component", () => {
    const { container } = render(
      <ExpandCollapse headerComponentRenderer={() => <h1>Some Header</h1>}>
        <h1>Child 1</h1>
        <span>Child 2</span>
      </ExpandCollapse>
    );

    expect(screen.getByText("Some Header")).toBeInTheDocument();

    expect(screen.queryByText("Child 1")).toBeNull();
    expect(screen.queryByText("Child 2")).toBeNull();

    fireEvent.click(container.firstChild.firstChild.firstChild);

    expect(screen.getByText("Child 1")).toBeInTheDocument();
    expect(screen.getByText("Child 2")).toBeInTheDocument();
  });
});
