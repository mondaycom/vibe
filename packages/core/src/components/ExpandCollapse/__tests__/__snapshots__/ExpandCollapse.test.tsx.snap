// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExpandCollapse > renders correctly with empty props 1`] = `
<div
  data-testid="expand-collapse"
  id=""
>
  <div
    className="expandCollapse showBorder"
  >
    <button
      aria-controls="-controls"
      aria-expanded={false}
      className="header section"
      onClickCapture={[Function]}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="icon iconComponent animateIconClose"
        data-testid="icon"
        data-vibe="Icon"
        fill="currentColor"
        height="24"
        viewBox="0 0 20 20"
        width="24"
      >
        <path
          d="M9.442 12.76a.77.77 0 0 0 1.116 0l4.21-4.363a.84.84 0 0 0 0-1.157.77.77 0 0 0-1.116 0L10 11.025 6.348 7.24a.77.77 0 0 0-1.117 0 .84.84 0 0 0 0 1.157l4.21 4.363Z"
        />
      </svg>
    </button>
  </div>
</div>
`;
